### PR TITLE
Fixed join on mixed nullability columns

### LINKED ISSUE #7936: [BUG] Hash join on structs throw exception caused by flattening null mask
**Describe the bug**
If probe table contains a non-null struct column and build table contains a nullable corresponding struct column, the number of columns of probe and build table will be different after `structs::detail::flatten_nested_columns`. Because build table contains extra column to represent validity of struct. The difference will fail the assertion at `cudf/cpp/src/join/hash_join.cu:365` with message  `Mismatch in number of columns to be joined on`.

**Steps/Code to reproduce bug**
```c++
auto int_col_a  = column_wrapper<int32_t>{{48, 27, 351, 31, 25}};
auto struct_col_a = cudf::test::structs_column_wrapper{{int_col_a}};
auto int_col_b  = column_wrapper<int32_t>{{351, 25, 27, 31, 48}};
auto struct_col_b = cudf::test::structs_column_wrapper{{int_col_b}, {1, 1, 0, 1, 1}};
std::vector<std::unique_ptr<cudf::column>> cols_a, cols_b;
cols_a.push_back(struct_col_a.release());
cols_b.push_back(struct_col_b.release());
Table ta(std::move(cols_a));
Table tb(std::move(cols_b));
auto result = cudf::inner_join(ta, t1, {0}, {0});
```

### PR DESCRIPTION
**Pull request by hyperbolic2346**
Fixing issue with join on a column with no validity joined with a column that had validity.

Also found and fixed an issue where the created nullability columns were not being retained throughout the operation resulting in undefined behavior.

Adjusted one of the existing tests to also test this issue.

Fixes #7936

### CODE REVIEW

File: cpp/src/join/hash_join.cu [hunk 1]
```diff
@@ -359,7 +362,8 @@ hash_join::hash_join_impl::compute_hash_join(cudf::table_view const &probe,
   CUDF_EXPECTS(probe.num_rows() < cudf::detail::MAX_JOIN_SIZE,
                "Probe column size is too big for hash join");
 
-  auto const _probe = std::get<0>(structs::detail::flatten_nested_columns(probe, {}, {}));
+  auto flattened_probe = structs::detail::flatten_nested_columns(probe, {}, {}, true);
+  auto const _probe    = std::get<0>(flattened_probe);
```
Comment (ttnghia) [inline cpp/src/join/hash_join.cu:366]: Do not use `_` prefix for local variables.

File: cpp/src/structs/utilities.cu [hunk 2]
```diff
@@ -86,9 +91,13 @@ struct flattened_table {
...
+      // nullable columns could be required for comparisions such as joins
       validity_as_column.push_back(cudf::is_valid(col));
-      validity_as_column.back()->set_null_mask(copy_bitmask(col));
+      if (col.nullable()) {
+        // copy bitmask only works if the column is nullable
+        validity_as_column.back()->set_null_mask(copy_bitmask(col));
+      }
       flat_columns.push_back(validity_as_column.back()->view());
       if (not column_order.empty()) flat_column_order.push_back(col_order);  // doesn't matter.
       if (not null_precedence.empty()) flat_null_precedence.push_back(col_null_order);
```
Comment (ttnghia) [inline cpp/src/structs/utilities.cu:103]:
```suggestion
      if (not column_order.empty()) { flat_column_order.push_back(col_order);  } // doesn't matter.
      if (not null_precedence.empty()) { flat_null_precedence.push_back(col_null_order); }
```

**Review by ttnghia — CHANGES_REQUESTED** [review]

**Review by ttnghia — APPROVED** [review]

File: cpp/src/structs/utilities.hpp [hunk 3]
```diff
@@ -66,7 +68,8 @@ std::tuple<table_view,
            std::vector<std::unique_ptr<column>>>
 flatten_nested_columns(table_view const& input,
                        std::vector<order> const& column_order,
-                       std::vector<null_order> const& null_precedence);
+                       std::vector<null_order> const& null_precedence,
+                       bool force_nullability_columns = false);
```
Comment (codereport) [inline cpp/src/structs/utilities.hpp:72]: Scoped enumerator for this? Maybe one already exists, i know there are a ton of `null` options

**Review by codereport — CHANGES_REQUESTED** [review]

File: cpp/src/structs/utilities.hpp (hunk 3 above)
Comment (ttnghia) [reply cpp/src/structs/utilities.hpp:72]: I vote for adding an enum to this.

File: cpp/src/structs/utilities.hpp [hunk 4]
```diff
@@ -23,6 +23,13 @@ namespace cudf {
 namespace structs {
 namespace detail {
 
+enum class column_nullability {
+  // generate a null column if the incoming column has nulls
+  MATCH_INCOMING,
+  // always generate a null column
+  FORCE
+};
+
```
Comment (codereport) [inline cpp/src/structs/utilities.hpp:32]: Just a suggestion
```suggestion
enum class column_nullability {
  MATCH_INCOMING, // generate a null column if the incoming column has nulls
  FORCE           // always generate a null column
};

```

**Review by codereport — APPROVED** [review]
LGTM, just one small suggestion

**Comment by hyperbolic2346** [comment]
@gpucibot merge

**Pull request merged**